### PR TITLE
Add Pascal Context to mmsegmentation

### PR DESCRIPTION
**Pull request by yamengxi**

### CODE REVIEW

File: mmseg/__init__.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 from .version import __version__, version_info
 
-MMCV_MIN = '1.1.2'
+MMCV_MIN = '1.1.0'
```
Comment (xvjiarui) [inline mmseg/__init__.py:5]: We may keep this file unchanged.

File: mmseg/datasets/pascal_context.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        'bed', 'bench', 'book', 'building', 'cabinet', 'ceiling', 'cloth',
+        'computer', 'cup', 'door', 'fence', 'floor', 'flower', 'food',
+        'grass', 'ground', 'keyboard', 'light', 'mountain', 'mouse',
+        'curtain', 'platform', 'sign', 'plate', 'road', 'rock', 'shelves',
+        'sidewalk', 'sky', 'snow', 'bedclothes', 'track', 'tree', 'truck',
+        'wall', 'water', 'window', 'wood')
+
+    '''
+    PALETTE is testing
+    '''
```
Comment (xvjiarui) [inline mmseg/datasets/pascal_context.py:25]: We may remove this line if the palette works fine.

File: mmseg/models/segmentors/encoder_decoder.py [hunk 3]
```diff
@@ -91,7 +91,7 @@ def encode_decode(self, img, img_metas):
             size=img.shape[2:],
             mode='bilinear',
             align_corners=self.align_corners)
-        return out
+        return out.exp()
```
Comment (xvjiarui) [inline mmseg/models/segmentors/encoder_decoder.py:94]:
```suggestion
        return out
```

File: mmseg/models/segmentors/encoder_decoder.py [hunk 4]
```diff
@@ -172,8 +172,8 @@ def slide_inference(self, img, img_meta, rescale):
         h_stride, w_stride = self.test_cfg.stride
         h_crop, w_crop = self.test_cfg.crop_size
         batch_size, _, h_img, w_img = img.size()
-        assert h_crop <= h_img and w_crop <= w_img, (
-            'crop size should not greater than image size')
+        # assert h_crop <= h_img and w_crop <= w_img, (
+        #     'crop size should not greater than image size')
```
Comment (xvjiarui) [inline mmseg/models/segmentors/encoder_decoder.py:176]: These lines could be removed. We may also add comments about what would happen if `h_crop > h_img`.

File: run.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
+GPUS=8 GPUS_PER_NODE=8 CPUS_PER_TASK=1 tools/slurm_train.sh mediaf pr_test configs/hrnet/fcn_hr48_480x480_40k_pascal_context.py --seed=0
```
Comment (xvjiarui) [inline run.sh:1]: This file is not necessary.

File: tools/convert_datasets/pascal_context.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
+import argparse
```
Comment (xvjiarui) [inline tools/convert_datasets/pascal_context.py:1]: We may add doc [here](https://github.com/open-mmlab/mmsegmentation/blob/master/docs/getting_started.md) about how to prepare pascal context dataset.

File: tools/slurm_test.sh [hunk 7]
```diff
@@ -21,4 +21,4 @@ srun -p ${PARTITION} \
     --cpus-per-task=${CPUS_PER_TASK} \
     --kill-on-bad-exit=1 \
     ${SRUN_ARGS} \
-    python -u tools/test.py ${CONFIG} ${CHECKPOINT} --launcher="slurm" ${PY_ARGS}
+    python -u tools/test.py ${CONFIG} ${CHECKPOINT} --launcher="slurm" ${PY_ARGS} &
```
Comment (xvjiarui) [inline tools/slurm_test.sh:24]: Unnecessary modification.

File: tools/slurm_train.sh [hunk 8]
```diff
@@ -20,4 +20,4 @@ srun -p ${PARTITION} \
     --cpus-per-task=${CPUS_PER_TASK} \
     --kill-on-bad-exit=1 \
     ${SRUN_ARGS} \
-    python -u tools/train.py ${CONFIG} --launcher="slurm" ${PY_ARGS}
+    python -u tools/train.py ${CONFIG} --launcher="slurm" ${PY_ARGS} &
```
Comment (xvjiarui) [inline tools/slurm_train.sh:23]: Unnecessary modification.

File: tools/test.py [hunk 9]
```diff
@@ -112,8 +112,15 @@ def main():
...
-    model.PALETTE = checkpoint['meta']['PALETTE']
+    if 'CLASSES' not in checkpoint['meta']:
+        model.CLASSES = dataset.CLASSES
+    else:
+        model.CLASSES = checkpoint['meta']['CLASSES']
+
+    if 'PALETTE' not in checkpoint['meta']:
+        model.PALETTE = dataset.PALETTE
+    else:
+        model.PALETTE = checkpoint['meta']['PALETTE']
```
Comment (xvjiarui) [inline tools/test.py:123]: Unnecessary modification.

File: configs/_base_/schedules/schedule_40k_lr4e-3.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+# optimizer
+optimizer = dict(type='SGD', lr=0.004, momentum=0.9, weight_decay=0.0001)
+optimizer_config = dict()
+# learning policy
+lr_config = dict(policy='poly', power=0.9, min_lr=1e-4, by_epoch=False)
+# runtime settings
+total_iters = 40000
+checkpoint_config = dict(by_epoch=False, interval=4000)
+evaluation = dict(interval=4000, metric='mIoU')
```
Comment (xvjiarui) [inline configs/_base_/schedules/schedule_40k_lr4e-3.py:9]: We may delete this file and modify the corresponding lr directly.

File: configs/_base_/schedules/schedule_80k_lr4e-3.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# optimizer
+optimizer = dict(type='SGD', lr=0.004, momentum=0.9, weight_decay=0.0001)
+optimizer_config = dict()
+# learning policy
+lr_config = dict(policy='poly', power=0.9, min_lr=1e-4, by_epoch=False)
```
Comment (xvjiarui) [inline configs/_base_/schedules/schedule_80k_lr4e-3.py:5]: We may delete this file and modify the corresponding lr directly.

**Comment by yamengxi** [comment]
I have resolved all the problems, please review.

File: mmseg/models/segmentors/encoder_decoder.py [hunk 12]
```diff
@@ -167,13 +167,15 @@ def forward_train(self, img, img_metas, gt_semantic_seg):
 
     # TODO refactor
     def slide_inference(self, img, img_meta, rescale):
-        """Inference by sliding-window with overlap."""
+        """Inference by sliding-window with overlap.
+
+        If h_crop > h_img or w_crop > w_img, we will not pad the crop_img but
+        send a smaller crop_img to encode_decode.
```
Comment (xvjiarui) [inline mmseg/models/segmentors/encoder_decoder.py:173]:
```suggestion
        If h_crop > h_img or w_crop > w_img, the small patch will be used to decode without padding.
```

Comment (yamengxi) [reply mmseg/models/segmentors/encoder_decoder.py]: It's ok, and I have committed it.

**Comment by yamengxi** [comment]
Here is the update comment version.

File: configs/deeplabv3/README.md [hunk 13]
```diff
@@ -41,3 +41,9 @@ Note: `D-8` here corresponding to the output stride 8 setting for DeepLab series
...
+|-----------|----------|-----------|--------:|----------|----------------|------:|--------------:|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| DeepLabV3 | R-101-D8 | 480x480   |   40000 |      9.2 |           7.09 | 46.55 |         47.81 | [model](https://openmmlab.oss-accelerate.aliyuncs.com/mmsegmentation/v0.5/deeplabv3/deeplabv3_r101-d8_480x480_80k_pascal_context/deeplabv3_r101-d8_480x480_40k_pascal_context_20200911_204118-1aa27336.pth) &#124; [log](https://openmmlab.oss-accelerate.aliyuncs.com/mmsegmentation/v0.5/deeplabv3/deeplabv3_r101-d8_480x480_80k_pascal_context/deeplabv3_r101-d8_480x480_40k_pascal_context-20200911_204118.log.json) |
```
Comment (xvjiarui) [inline configs/deeplabv3/README.md:48]: These filenames are not the same as the paths provided. Please check.

Comment (xvjiarui) [reply configs/deeplabv3/README.md:48]: Should be `pascal_context-` not `pascal_context_`.

Comment (xvjiarui) [reply configs/deeplabv3/README.md:48]: Also, check the timestamp.

Comment (yamengxi) [reply configs/deeplabv3/README.md:48]: you means should be `pascal_context_` not` pascal_context-`

**Comment by yamengxi** [comment]
I have fixed the model path mistakes.

**Review by xvjiarui — APPROVED** [review]

**Comment by hellock** [comment]
Task linked: [CU-4fp4de Pascal Context](https://app.clickup.com/t/4fp4de)

**Comment by OpenMMLab-Assistant-007** [comment]
Hi!
@yamengxi 
First of all, we want to express our gratitude for your significant PR in the OpenMMLab project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**